src/Storages/MergeTree/MergeTreeIndexText.cpp

```diff
@@ -280,22 +270,62 @@ ColumnPtr deserializeTokensFrontCoding(ReadBuffer & istr, size_t num_tokens)
 
     return tokens_column;
 }
-}
 
 /// TODO: add cache for dictionary sparse index
-void MergeTreeIndexGranuleText::deserializeSparseIndex(ReadBuffer & istr)
+DictionarySparseIndex deserializeSparseIndex(ReadBuffer & istr)
 {
     ProfileEvents::increment(ProfileEvents::TextIndexReadSparseIndexBlocks);
 
     size_t num_sparse_index_tokens = 0;
     readVarUInt(num_sparse_index_tokens, istr);
 
+    DictionarySparseIndex sparse_index;
     sparse_index.tokens = deserializeTokensRaw(istr, num_sparse_index_tokens);
 
     auto offsets_in_file = ColumnUInt64::create();
     SerializationNumber<UInt64> serialization_number;
     serialization_number.deserializeBinaryBulk(*offsets_in_file, istr, 0, num_sparse_index_tokens, 0.0);
     sparse_index.offsets_in_file = std::move(offsets_in_file);
+
+    return sparse_index;
+}
+
+BloomFilter deserializeBloomFilter(ReadBuffer & istr, const MergeTreeIndexTextParams & params, size_t num_tokens)
+{
+    size_t bytes_size = getBloomFilterSizeInBytes(params.bloom_filter_bits_per_row, num_tokens);
+
+    BloomFilter bloom_filter(bytes_size, params.bloom_filter_num_hashes, 0);
+    istr.readStrict(reinterpret_cast<char *>(bloom_filter.getFilter().data()), bytes_size);
+
+    return bloom_filter;
+}
+}
+
+void MergeTreeIndexGranuleText::deserializeHeader(ReadBuffer & istr, const MergeTreeIndexDeserializationState & state)
+{
+    const auto & condition_text = typeid_cast<const MergeTreeIndexConditionText &>(*state.condition);
+
+    /// Either retrieves a text index header from cache or from disk when cache is disabled.
+    const auto get_header = [&]() -> TextIndexHeaderPtr
+    {
+        const auto load_header = [&]
+        {
+            size_t num_tokens;
+            readVarUInt(num_tokens, istr);
+            auto bloom_filter = deserializeBloomFilter(istr, params, num_tokens);
+            auto sparse_index = deserializeSparseIndex(istr);
+            return std::make_shared<TextIndexHeader>(num_tokens, std::move(bloom_filter), std::move(sparse_index));
+        };
+
+        if (condition_text.useHeaderCache())
+            return condition_text.headerCache()->getOrSet(
+                TextIndexHeaderCache::hash(state.path_to_data_part, state.index_name, state.index_mark),
+                load_header);
+
+        return load_header();
+    };
```

Comment:
I think that the outer lambda is not needed, just an if-else could do the same much simpler.

Reply:
in L#323 the `load_header` function is passed to the cache interface but it's only executed when an entry is not in cache. So, like that I tried to avoid duplication with if/else